### ISSUE TITLE
#[rpc] macro in derive has dependencies which are not propagated into consuming projects

### ISSUE THREAD

**Issue by fwcd** [paritytech/jsonrpc#481]
I am trying to implement the [basic example using derive](https://github.com/paritytech/jsonrpc#basic-usage-with-derive) from the README. Since the code snippet only seems to require `jsonrpc-core` and `jsonrpc-derive`, I did not include any other dependencies in `Cargo.toml`:

```toml
[dependencies]
jsonrpc-core = "13.0.0"
jsonrpc-derive = "13.0.0"
```

The compiler, however, yields an error that a crate named `serde` is required too:

```
error: Could not find `serde` in `dependencies` or `dev-dependencies` in `...\Cargo.toml`!
 --> src\main.rs:4:1
  |
4 | #[rpc]
  | ^^^^^^
```

Indeed does the `#[rpc]` macro have a dependency on `serde`:

https://github.com/paritytech/jsonrpc/blob/54c4de527f536f8b6c5c6f4616446f82d00ed936/derive/src/rpc_trait.rs#L258-L261

`serde` is only listed as a dev-dependency though, so it [will not propagate](https://doc.rust-lang.org/rust-by-example/testing/dev_dependencies.html) into projects consuming the library:

https://github.com/paritytech/jsonrpc/blob/54c4de527f536f8b6c5c6f4616446f82d00ed936/derive/Cargo.toml#L21-L30

Adding `serde` explicitly to my Cargo.toml did fix the issue, but brought up another missing dependency (`jsonrpc-core-client`) which, again, had to be added manually despite being used by the `#[rpc]` macro.

Am I missing something? Would it be possible to clarify which dependencies have to be provided manually when using `jsonrpc-derive`?

**Comment by dvdplm** [comment]
This sounds like a bug to me. Want to attempt a PR to fix it? :)

**Comment by tomusdrw** [comment]
Yeah, so this is a tricky one. The dependencies can't be just in `jsonrpc-derive` crate - we would need to re-export them, because the macro generates code in the crate when it's used.

However the set of crates that is needed depends on the features you're going to use. e.g.` jsonrpc-core-client` is only necessary if the client code is being generated.

I'm ok with re-exporting deps that are 100% required (like `serde`), but have mixed feelings about `jsonrpc-core-client` for instance.

**Comment by simeir** [comment]
> Yeah, so this is a tricky one. The dependencies can't be just in `jsonrpc-derive` crate - we would need to re-export them, because the macro generates code in the crate when it's used.
> 
> However the set of crates that is needed depends on the features you're going to use. e.g.` jsonrpc-core-client` is only necessary if the client code is being generated.
> 
> I'm ok with re-exporting deps that are 100% required (like `serde`), but have mixed feelings about `jsonrpc-core-client` for instance.

Whether the client crate is needed is based on the option provided for the attribute. (https://github.com/paritytech/jsonrpc/blob/9360dc86e9c02e65e858ee7816c5ce6e04f18aef/derive/src/options.rs#L17)

I just did an experiment. If the attribute is `#[rpc(server)]`, it's not necessary to add `jsonrpc-core-client`. if `#[rpc]` or `#[rpc(client)]` is used, it must be there, which makes sense, since I can see that `#[rpc]` is designed to use both client and server functionalities, and `#[rpc(client)]` apparently needs stuff written for the client.

So I think we will be able to solve this problem by, to re-export `serde`, and to have better documentation and examples.

**Comment by simeir** [comment]
I just created a pull request for this (#498 ).
